Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,61 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
```

Comment:
I don't understand this special case. Is it needed?

Reply:
Yes, as a base case for the recursion and to build a key even for empty tuples.